// nad_circapi/index.js
var api         = require('circonusapi2'),
    os          = require('os'),
    fs          = require('fs'),
    exec        = require('child_process').exec,
    target      = null,
    hostname    = null,
    brokerid    = null,
    configfile  = null,
    config      = null;

var required_graph_data_fields = [
    { "name": "stack",          "default": null },
    { "name": "hidden",         "default": false },
    { "name": "name",           "default": null },
    { "name": "axis",           "default": "l" },
    { "name": "color",          "default": null },
    { "name": "alpha",          "default": 0 },
    { "name": "derive",         "default": "gauge" },
    { "name": "legend_formula", "default": null },
    { "name": "data_formula",   "default": null }
];

var required_composite_data_fields = [
    { "name": "stack",          "default": null },
    { "name": "hidden",         "default": false },
    { "name": "name",           "default": null },
    { "name": "axis",           "default": "l" },
    { "name": "color",          "default": null },
    { "name": "legend_formula", "default": null },
    { "name": "data_formula",   "default": null }
]

exports.configure = function (tok, targ, host, brok, conf) {
    target      = targ;
    hostname    = host;
    brokerid    = brok;
    configfile  = conf;

    api.setup(tok, 'nad');

    if ( hostname === null ) {
        lookup_hostname();
    }
    else {
        get_config();
    }
};

function lookup_hostname() {
    if ( os.type() === 'SunOS' ) {
        exec("/usr/bin/zonename", function(error, stdout, stderr) {
            if ( error != null ) {
                console.error("Could not look up hostname: " + error);
                process.exit(1);
            }
            hostname = stdout.replace(/\n/g, "");
            get_config();
        });
    }
    else {
        console.log("don't know what I am supposed to look up");
    }
}

function get_config() {
    try {
        config = fs.readFileSync(configfile).toString();
        config = JSON.parse(config);
    }
    catch (err) {
        console.error("Could not read config ["+configfile+"] " + err);
        process.exit(1);
    }

    configure_check();
}

function configure_check() {
    var metrics = [];
    for ( var idx in config.metrics.numeric ) {
        metrics.push({"name": config.metrics.numeric[idx], "type": "numeric"});
    }    
    for ( var idx in config.metrics.text ) {
        metrics.push({"name": config.metrics.text[idx], "type": "text"});
    }

    // Create the check via the Circonus API
    api.post(
        "/check_bundle",
        {
            "target": target,
            "type": "json",
            "display_name": hostname + " nad",
            "brokers": [ brokerid ],
            "period": 60,
            "timeout": 10,
            "metrics": metrics,
            "config": {
                "port": 2609,
                "url": "http://"+target+"/"
            }
        },
        function(code, error, body) {
            if ( code === 401 ) {
                console.error("The node agent is not permitted to talk to the API with this auth token, please log in to Circonus and approve nad as an app");
                process.exit(1);
            }
            else if ( code !== 200 ) {
                console.error("Error from Circonus API: " + error);
                process.exit(1);
            }
            else {
                var lookup = {};
                // We are going to compare the metrics we got back from the API with what we wanted.  The API check_bundle endpoint
                // does some deduplication on check creation and we might have been handed back an already created check which is 
                // just missing some additional metrics we want.
                for ( var idx in body.metrics ) {
                    if ( ! lookup[body.metrics[idx].type] ) {
                        lookup[body.metrics[idx].type] = {};
                    }
                    lookup[body.metrics[idx].type][body.metrics[idx].name] = 1;
                }

                var needs_update = 0;
                for ( var idx in metrics ) {
                    if ( ! lookup[metrics[idx].type] || ! lookup[metrics[idx].type][metrics[idx].name] ) {
                        body.metrics.push(metrics[idx]);
                        needs_update = 1;
                    }
                }

                // If we were missing any metrics in the return, we will need to update our existing check.  The missing metrics were
                // added to the body, so PUT that back at the API using the object's _cid as the endpoint
                if ( needs_update ) {
                    api.put(
                        body._cid,
                        body,
                        function(code, error, body) {
                            if ( code !== 200 ) {
                                console.error("Error from Circonus API: " + error);
                                process.exit(1);
                            }
                            configure_graphs(body._checks[0].split("/")[2]);
                        }
                    );
                }
                else {
                    configure_graphs(body._checks[0].split("/")[2]);
                }
            }
        }
    );
}

function configure_graphs(check_id) {
    for ( var name in config.graphs ) {
        var graph = config.graphs[name];
        graph.title = hostname + " " + name;

        for ( var idx in graph.datapoints ) {
            graph.datapoints[idx].check_id = check_id;

            for ( var iidx in required_graph_data_fields ) {
                if ( ! graph.datapoints[idx][required_graph_data_fields[iidx].name] ) {
                    graph.datapoints[idx][required_graph_data_fields[iidx].name] = required_graph_data_fields[iidx].default;
                }
            }
        }

        if ( graph.composites ) {
            for ( var idx in graph.composites ) {
                for ( var iidx in required_composite_data_fields ) {
                    if ( ! graph.composites[idx][required_composite_data_fields[iidx].name] ) {
                        graph.composites[idx][required_composite_data_fields[iidx].name] = required_composite_data_fields[iidx].default;
                    }
                }
            }
        }

        post_graph(graph);
    }
}

function post_graph(graph) {
    api.post(
        "/graph",
        graph,
        function(code, error, body) {
            if ( code !== 200 ) {
                console.error("Could not create graph " + graph.title, error);
            }
        }
    );
}
